### DISCUSSION TITLE
New distribution type: source

### DISCUSSION THREAD

**Discussion by loicrouchon** [jreleaser/jreleaser#1154] [Ideas]
Hi!

I'm experimenting with the creation of source packages at the moment with the Spec packager.
I would also like in the future to provide debian source packages (I know there is no such packager at the moment) and also ideally provide a source based brew Formula.

**Why?**

Many package managers only accept software that can be built from sources in their official repositories. Even if they usually provide a way to support external repositories where binary packages can be stored, providing only a binary package would prevent any inclusion in the official repositories.

**What I did so far**

At the moment, jreleaser does not support source distributions, but I was able to play a bit with the Spec packager to trick it.

I did override the app.spec.tpl to include the `BuildRequires` as well as the proper build instructions so that it can be built by https://copr.fedorainfracloud.org for the supported distributions.

_**The only thing that is annoying is that I had to provide an artifact.**_

**A potential solution**

I haven't found at the moment a satisfactory solution to this with jreleaser.

I only see two options at the moment:

1. Not using jreleaser for source distributions

This would imply:

* generating the spec file as part of my build process without leveraging what jreleaser knows about my project (Name, License, ...)
* Create a jreleaser distribution referencing the spec file. But here, not sure what distribution type should be used as I get a `[WARN]  distribution.rpmSpec.spec was disabled because there are no matching artifacts [.tar, .tar.xz, .zip, .tar.gz, .tar.bz2, .tbz2, .txz, .tgz]` when not specifying the distribution type.

Or

2. Introduce a new distribution type `SOURCE` which requires 0 artifacts.

Of course, the packager recipe (Spec file, Formula, ...) would need to be adapted to support build dependencies as well as build commands and other required information. But this can be mitigated in the beginning by overriding those recipes

So I would like to know the direction that jreleaser will be taking regarding support for source distributions:

1. Support it?
2. Not supporting it?

Cheers!

**Comment by aalmiray** [comment]
@loicrouchon Thank you for raising this topic. The original idea for JReleaser was to make it easier to publish JavaFX applications to Homebrew, hence the "tunnel vision" into providing support for binary releases. This being said, the tool has evolved a lot since then. I do like the idea of `SOURCE` distributions. The catch is figuring out a couple of potential consumers; at this point we have `rpm` and `deb`, but there could be others. Perhaps even branch out into PyPy, Ruby gems, and additional targets where source is preferred.

Would a single `SOURCE` distribution work? I'm not so sure yet given that we have several binary distributions due to constraints applied by package managers to each one of them. Could it be that a `SOURCE` distribution requires:
- a set of FileSet, one per each source directory.
- a set of dependencies in which case the format is specific to the source language _or_ target package manager, or both.

Additionally, I've also think that opening the door to generating arbitrary files via templates could help (see https://github.com/jreleaser/jreleaser/discussions/969). For example, one could generate a spec file based on the data provided by `jreleaser.yml` and publish it via an HTTP uploader, or release it as a Git release asset. The next step would be pushing it to the next node in the chain such as Fedora COPR.

Perhaps both options would help attain the goal you seek. What I would propose is to continue the discussion fleshing out potential DSL updates with `rpm` and `deb` to begin with. WDYT?

**Comment by loicrouchon** [comment]
@aalmiray thanks for your quick reply:

> The catch is figuring out a couple of potential consumers; at this point, we have rpm and deb, but there could be others.

If I'm not mistaken, the default way of doing things with `brew` is to build from source and have a CI build the bottles. So the current implementation of `brew` in jreleaser cannot be included in the official `brew` repository (It would also require a PR publisher on top of the `SOURCE` distribution).
So yes, `deb`, `rpm`, (`PyPy`, `Ruby gems` no experience there so I trust you), but also `brew`.

> Could it be that a SOURCE distribution requires:
>  * a set of FileSet, one per each source directory.

I am not sure about the approach here. On my project, I wrote the [RPM Spec file](https://github.com/loicrouchon/symly/blob/main/src/packaging/fedora/symly.spec) to reference the GitHub tagged sources. Such URLs are of the form `https://github.com/<project>/<repository>/archive/refs/tags/v<version>.tar.gz`

I did so because it was quite painful to define what files to include in that tar:

* Everything minus the content of the `.gitignore`? 
* Only what is needed?
* Only the content of `src`?

But I do agree that some package managers won't accept a reference to an URL having the source code. If I recall correctly, it is not possible for Debian.

> Could it be that a SOURCE distribution requires:
>  * a set of dependencies in which case the format is specific to the source language or target package manager, or both.

I think we need more than this. For example in this [RPM Spec file](https://github.com/loicrouchon/symly/blob/main/src/packaging/fedora/symly.spec) (not a state-of-the-art spec file, but good enough for a start), I had to define:

* The build dependencies (`BuildRequires`)
* The runtime dependencies (`Requires`)
* The build command (`%build`)
* The installation commands (`%install`): I think hard to be determined automatically
* The files that will be installed (`%files`): I think hard to be determined automatically

But I do believe it is not necessary to go that deep in the first attempt. The issue https://github.com/jreleaser/jreleaser/discussions/969 you mentioned about templating should help to provide a great baseline to start with.

> Perhaps both options would help attain the goal you seek. What I would propose is to continue the discussion fleshing out potential DSL updates with rpm and deb to begin with. WDYT?

I'm definitely up for continuing this discussion.

`rpm` + `deb` is a good way to start with. For the `rpm` part I already managed to get a spec. For the `deb` part, I haven't managed yet to create a proper source package, so I'll need to dig more here to get some experience first.

But we should include at least those two so that we have a better overview of the different requirements a source-based packaging can impose.

**Reply by aalmiray** [reply]
> If I'm not mistaken, the default way of doing things with brew is to build from source and have a CI build the bottles. So the current implementation of brew in jreleaser cannot be included in the official brew repository (It would also require a PR publisher on top of the SOURCE distribution).

True, that's why JReleaser currently supports publishing to your own tap. This being said it seems there may be an a way in for Java tools to be added to homebrew-core, as both JReleaser and JBang have been added before New Year's eve in 2022. This of course doesn't automatically apply for non Java projects.

> I am not sure about the approach here. On my project, I wrote the [RPM Spec file](https://github.com/loicrouchon/symly/blob/main/src/packaging/fedora/symly.spec) to reference the GitHub tagged sources. Such URLs are of the form https://github.com/<project>/<repository>/archive/refs/tags/v<version>.tar.gz

A repository tarball (like the one you linked) would likely suffice for many consumers. In the case that it's not a source archive (`tar.gz`, `zip`, etc) could be uploaded at a particular place where the packaging service expects it.

> I think we need more than this. For example in this [RPM Spec file](https://github.com/loicrouchon/symly/blob/main/src/packaging/fedora/symly.spec) ...

I think we're mixing distribution with package manager. A distribution defines metadata about artifacts that may be consumed by package managers to generate required files for publication. In the case of the `SOURCE` distribution these artifacts are the source trees (if more than one git repo perhaps?). A package manager such as `RPM` or `DEB` may consume distributions of type `SOURCE`, adding whatever required properties for each, such as dependencies, maintainer, etc. There may be common properties among package managers, in which case some may be moved to the `distribution`, or even all thew way to the `project`. Such was the case for `icons` and `screenshots` used by flatpak and appimage.
